src/contexts/addresses.tsx

```diff
@@ -319,7 +320,7 @@ export const AddressesContextProvider: FC<{ overrideContextValue?: PartialDeep<A
 
       const addressesMetadata = loadStoredAddressesMetadataOfAccount(currentAccountName)
 
-      if (addressesMetadata.length === 0) {
+      if (usedPassphrase || addressesMetadata.length === 0) {
```

Comment:
By removing this newly added condition we could have shared labels between decoy and passphrase wallets. Not sure we would want that though. However, we cannot proceed without finding a solution to the labeling issue I mentioned in my other comment.

Reply:
> Not sure we would want that though

Yeah; you know the implications.

I'll work on a solution for the labeling

Reply:
This is a comment on what's been said so far and my thoughts:

> in my decoy wallet I had set the label Vault [...] 

So yes, this is the correct behavior, for labels not to be saved, since it would void plausible deniability...

This is related again to needing to store information about password derived wallets... As soon as we store anything, it opens the doors to revealing other wallets exist.

After some time has passed since the initial implementation, I'm...willing to accept to void plausible deniability for a friendlier UX, which would include the "this passphrase has no associated wallet" message. What I am happy about is at its core, we still follow BIP39, it's just augmented with some things which void some of its benefits... /cc @polto @mvaivre if you're ok with this.

> I have 2 main addresses.

This is a bug / unexpected though... There should only be the main address associated with the non passphrase derived wallet

Reply:
Yeah, labels (and any other metadata) would void plausible deniability more clearly than the message we've debated a lot about ("this passphrase has no associated wallet") 🙃

Encrypting the metadata file would solve the issue though. Thinking out loud : we need to keep in mind that one day we may want to have a service to share metadata between multiple wallets, so we shouldn't encrypt the file using the device specific credentials (password, pin or biometrics). We could maybe use the X first words of the mnemonic?

User authenticates => wallet is decrypted => first X words of mnemonic are used to decrypt the metadata file => wallet is operational.

WDYT?

Reply:
The argument for losing the plausible deniability feature was not the message "this passphrase has no associated wallet", but the fact that we are storing an encrypted entry in localStorage. @polto and @LeeAlephium pointed out that this would cause the advanced attacker to think that this is the passphrase, meaning they would then torture the victim to reveal it, even if they didn't set one. I raised the issue that _"Does this mean we can never have anything encrypted stored in localStorage"_ to which no one really properly replied.

Reply:
Mhh ok, this is not what I understood, but I was not involved deeply in the discussion. I'm hence probably wrong. For me, here's the typical scenario where you want to deny that you have access to a big wallet : 

1) The attacker wants you to open your wallet(s) or you die.
2) You show the attacker the decoy wallet, with some coins in it. 
3) He/she says: but heyyy, I'm clever and I see that you have another encrypted file there! You used a passphrase! Open it!
4) You say, "oh yeah, forgot about this, I'm sorry... I experimented with it in the past, but I don't think I have anything on it!". You type any passphrase, it opens an empty wallet. "You see?". 
5) "Well okay then." he / she says. "Have a nice day!". ☀️  ... OR they kill you anyway, because they understand the tech deeply and they don't trust you. But this is not something we can fix.

This works only if we don't show the "this passphrase has no associated wallet" message.

Anyway : if we store non-encrypted metadata, the scenario above is compromised as soon as the attacker looks at the localstorage (if possible?).

Reply:
There's no way to access devtools in the production build (I think they may even be entirely absent), so checking it that way is not possible. The only way is they bring their own debugging tools with them.

I think it's a good idea to continue to leave out the message.

> "Does this mean we can never have anything encrypted stored in localStorage" to which no one really properly replied.

No; we can have things stored encrypted, but a (sophisticated, highly focused) attacker will demand all data to be decrypted. A passphrase derived wallet leaves no trace behind, so there is still no way for the attacker to know it exists. When storing _anything related to a derived wallet_ though, that's what voids plausible deniability... Even 1 bit of information voids it... 

I'm being lenient on allowing labels since I'm in agreement with @nop33 that the feature should still be accessible to a large amount of users. 

I have come to accept that the desktop-wallet needs to remain friendly and not expose advanced features. There can be another wallet by someone else which offers it all. 

My comment is now leading into a tangent but we should re-evaluate the features the wallet offers based on this view of the wallet (things like sweeping are advanced IMO).

Reply:
I believe that we can find a balance, like we have done until now. IMO we successfully added convenient features for everyday users (labels etc.), while hiding advanced features efficiently (but still offering them). We can do better, but it's a nice start.

Let's not forget that we're at the beginning of the network, and we **need** to onboard advanced users to help us grow the network (devs, miners, Defi users etc.). 

So it's not a black or white / easy vs. complex situation. I'm sure we can find UI tricks to hide the complexity away.

I'll do a big redesign pass later on, as I also need to take care of Cheng's additions. I'll find ways to make this simple again.

Reply:
if you start with a fresh wallet and add a 12-24 BIP39 words wallet, you will have wallet data including the seed encrypted with wallets's password. Labels are probably stored encrypted with that same password.

if you add a passphrase we could have an "amnesic" mode. Or store the seed without the passphrase (potentially with its doublehash) and account labels encrypted with password+passphrase. 

We can have several slots for passphrase enabled wallets and we can by default create 10 with random encrypted data. Users can overwrite slots. This way there is no possibility to tell if it's the 10 random ones created by default or if the user has a password + passphrase valid for that slot. 

This way we preserve plausible deniability and have nice features such as labels.

But first we can start with amnesic wallet (in terms of labels and local data storage) because it's more simple.

Reply:
Might be easier to actually go ahead with the 10 random encrypted data... The wallet makes no special distinction between passphrase or non-passphrase derived wallets...

How is it known what's a considered as a wallet which can be replaced / is a "dummy" wallet? I think rather than slots, the wallets just exist in storage but arent listed in the wallet listing...

Reply:
I've put the PR back into WIP to do this later :)

Reply:
This is looking really damn good guys. I'm ultra happy we stuck with BIP39 and then augmented it.

I think there's no point in an arbitrary number of random wallets. I think it's better to stick to fundamentals and ask ourselves what it is that actually gives us plausible deniability here.

* 0 wallets: The user has no ALPH.
* 1 wallets: The user has ALPH.
* 1+n wallets: The user has ALPH but no way to know how much.

Based on this, I suggest we simply make sure there is always a single extra dummy wallet in storage. This means the user can even have 2 wallets and 1 dummy and it's impossible to know the user has more than 1.

Ready to be reviewed soon as of this comment...

Reply:
I finally got time to go through all discussions. Thank you @polto and @lf94 for the ideas.

I want to propose a solution that is somewhere in the middle between @polto's and @LeeAlephium's.

@polto's [idea](https://github.com/alephium/desktop-wallet/pull/244#discussion_r882482267) cannot really give us plausible deniability, because the system cannot know which of the "slots" contain random data and are available to be overridden and which actually contain useful metadata.

@LeeAlephium's [idea](https://github.com/alephium/desktop-wallet/pull/244#discussion_r884039650) of a dice roll creates an indefinite number of metadata in the localStorage and depends on how often the user uses the wallet. I don't like that, even if it provides 100% plausible deniability, but can create a lot of garbage along the way, which we will have to carry along.

My proposal for storing address metadata without sacrificing plausible deniability is:

## 1. address metadata storage key
Each wallet stores its metadata. The key is `'address-metadata-' + doublehash('my wallet name' + passphrase)` (same as @LeeAlephium's [proposal](https://github.com/alephium/desktop-wallet/pull/244#discussion_r881980540)). So, for a wallet called _"Ilias' badass wallet"_ with no passphrase, and for the same wallet, but with a passphrase, the keys of the 2 address metadata entries could look like this:
```
addresses-metadata-00f96c9b1db7a585d4ba3d2a4af32ca3f95d20ef5766e1773cf0c07e12672b6516e6f21c679f40c95cefdf41c1ce37724c6f3781caaabf0b481391a4144978ee
addresses-metadata-b2189fac31a86e96d637b492378fbda08f51bfebfa9e3b2f95f5d36246b75d85c2a9eea9e545c2bfd85dbf4d5b7deae4986dad6e18109c8842790d8bd695bc87
```

## 2. address metadata storage value
The value of the address metadata needs to be stored encrypted.

@polto proposed we use `password + passphrase` as the encryption key. This approach would work, but it would make the future transfer of metadata across devices problematic (@mvaivre's [request](https://github.com/alephium/desktop-wallet/pull/244#discussion_r881290829)).

@mvaivre proposed we use the `X` first words of the 24-words mnemonic. @LeeAlephium proposed to use the whole 24-words mnemonic. Both of these approaches suffer the same problem: The attacher knows the password to the 24-word mnemonic wallet, which means they know the 24-words, which means they can try and decrypt address-metadata, and if they find one where the decryption is successful, it would reveal that the user is using a passphrase.

To solve both problems mentioned above, I propose using the 24-words mnemonic + passphrase as the encryption key for the address metadata storage value.

## 3. Maintaining (most of) plausible deniability
Instead of 10 slots with random metadata which the wallet cannot know which it can override (@polto's idea), and instead of probabilistically generating metadata every time the user logs in (@LeeAlephium's idea), I propose that the wallet picks a random number `Y` between 5 and 20 (upper bound TBD), and it generates ONLY ONCE `Y` address metadata entries with garbage. So, a user with 3 wallets (_"My wallet"_, _"My mom's wallet"_, _"My sister's wallet"_) and 1 additional passphrase wallet will have 4 + `Y` entries. Plausible deniability is void only when that number exceeds the upper bound.

Each time a wallet is unlocked using a random passphrase, a new address metadata entry is created. To avoid cluttering the storage, we can avoid storing address metadata when there is only 1 address and when this address has no metadata associated with it (the default state of a new wallet).

Reply:
> We hadn't decided which approach to go for, we're still discussing.

It was more about it's problematic than reaching a conclusion

> Cons: not 100% plausible deniability

It's either 0% or 100%, there's no inbetween - I think a better description is "maximum amount of wallets until plausible deniability is voided"

Another problem with proposition 3 is their **entire** wallet transitions from 100% plausibly deniable to 0% plausibly deniable. That's a big transition...

> On the contrary. People where Y is close to the low bound have a lot of space for creating wallets. People where Y is close to the upper bound and have many wallets will have their plausible deniability voided (example: Y is 18 and the user has 2 wallets and 1 passphrase wallet, the total number of address metadata entries is 21, which is above the upper bound of the random range for Y, which is 20 on the above proposal).

Maybe you misunderstood my comment; but what's being written isn't adding up:

* Generate a number 5 >= Y <= 20
* Generate Y wallets
* Let there be X user generated wallets
* Plausible deniability voided if X > Y

If Y is 5 it will be voided much sooner than if it were 20?...

Reply:
@nop33 clarified offline: It should be if X + Y > UpperRange, then it's voided

So it's UpperRange - LowerRange which determines how many wallets the user can create until it's voided.